src/D2L.Security.OAuth2/Provisioning/Default/AuthServiceClient.cs

```diff
@@ -45,17 +57,72 @@ Uri authEndpoint
 		/// <param name="assertion">A JWT signed by the private key of the entity requesting the token</param>
 		/// <param name="scopes">List of scopes to include in the access token</param>
 		/// <returns>A JWT token from the auth service signed with the auth service's private key</returns>
+		/// <exception cref="AuthServiceException">
+		/// The auth service could not be reached, or it did not respond with
+		/// a status code indicating success.
+		/// </exception>
 		async Task<IAccessToken> IAuthServiceClient.ProvisionAccessTokenAsync(
 			string assertion,
 			IEnumerable<Scope> scopes
 		) {
 			string requestBody = BuildFormContents( assertion, scopes );
-			using( HttpResponseMessage response = await MakeRequest( requestBody ).SafeAsync() ) {
-				response.EnsureSuccessStatusCode();
-
-				using( var resultStream = await response.Content.ReadAsStreamAsync().SafeAsync() ) {
-					IAccessToken accessToken = SerializationHelper.ExtractAccessToken( resultStream );
-					return accessToken;
+			HttpResponseMessage response = null;
+			try {
+				try {
+					response = await MakeRequest( requestBody ).SafeAsync();
+				} catch( TaskCanceledException exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.Timeout,
+						message: "The web request to the Auth Service has timed out.",
+						innerException: exception
+					);
+				} catch( Exception exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ConnectionFailure,
+						message: "Could not establish a connection with the Auth Service.",
+						innerException: exception
+					);
+				}
+				
+				if( !response.IsSuccessStatusCode ) {
+					string errorMessage;
+					try {
+						string json = await response.Content.ReadAsStringAsync().SafeAsync();
+						if( string.IsNullOrWhiteSpace( json ) ) {
+							errorMessage = response.ReasonPhrase;
+						} else if( !SerializationHelper.TryExtractErrorMessage( json, out errorMessage ) ) {
+							errorMessage = string.Concat( response.ReasonPhrase, ": ", json );
+						}
+					} catch( Exception ) {
+						errorMessage = response.ReasonPhrase;
+					}
+					
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ErrorResponse,
+						serviceStatusCode: response.StatusCode,
+						message: errorMessage
+					);
+				}
+				
+				string responseJson = string.Empty;
+				if( response.Content != null ) {
+					responseJson = await response.Content.ReadAsStringAsync().SafeAsync();
+				}
+					
+				IAccessToken token;
+				if( SerializationHelper.TryExtractAccessToken( responseJson, out token ) ) {
+					return token;
+				}
+				
+				throw new AuthServiceException(
+					errorType: ServiceErrorType.ClientError,
+					message: string.IsNullOrEmpty( responseJson ) ?
+						EMPTY_RESPONSE_ERROR_MESSAGE :
+						string.Concat( INVALID_JSON_ERROR_MESSAGE_PREFIX, responseJson )
+				);
+			} finally {
+				if( response != null ) {

```

Comment:
`.SafeDispose()` should now be available if you rebase: https://github.com/Brightspace/D2L.Services.Core.Extensions/blob/master/src/D2L.Services.Core.Extensions/System.IDisposable.Extensions.cs#L8